### ISSUE TITLE
delete-then-download doesn't delete container with contract violation before download new container

### ISSUE THREAD

**Issue by ClintHaerinck** [balena-os/balena-supervisor#1865]
When for example doing an update between L4T versions of a service on a Jetson platform.

**Comment by pipex** [comment]
Hi @ClintHaerinck, could you provide more details on how you observed this behavior?

If a release has a service with a contract violation, there are only two things that should happen
- The service is set as optional, which means the supervisor will remove it from the local copy of the target release and not even download the image
- The service is not optional, meaning the supervisor will reject the install of the release completely, maintaining the current release.

Have you confirmed that the contract violating service has had its image downloaded and the container has been installed installed? Could you also share your device type, OS and supervisor versions?

Thank you.

**Comment by ClintHaerinck** [comment]
This is how it went:
* Had a device with L4T R32.4.4 (balenaOS 2.67.3+rev2), a release is deployed at balenaCloud with 2 optional services. One for L4T R32.4.4 and one for L4T R32.5.1.
* I've updated the device to L4T R32.5.1 (balenaOS 2.82.11+rev11). After a successful balenaOS update, the old (R32.4.4) service and matching image wasn't removed before downloading the new service (R32.5.1).

As result the device gets out of space and unbootable. Not sure about the supervisor versions, but the latest at that time.

**Comment by ClintHaerinck** [comment]
@acostach I've just tested this again with latest supervisor (v12.11.38).

The containers are deleted before the download of the new L4T image but the huge image (6GB, L4T R32.4.4) isn't deleted before the download of the new huge image (6GB, L4T R32.5.1).

**Comment by ClintHaerinck** [comment]
Ok, I've managed to use the functionality:

* Using the variable `RESIN_SUPERVISOR_UPDATE_STRATEGY` instead of `BALENA_SUPERVISOR_UPDATE_STRATEGY` at BalenaOS 2.67.3+rev2 . The name of the supervisor container is still `resin_supervisor` also.

Are there some HostOS dependencies to use `BALENA_*` variables?

The documentation at following page doesn't include above variables and the link to improve the docs doesn't work:
https://www.balena.io/docs/reference/supervisor/configuration-list/jn30b-nano/

However when upgrading the OS from 2.67.3+rev2 to 2.82.11+rev11, after the upgrade I still need to do an extra reboot to delete the image with contract violations or it isn't deleted before downloading the new one. This is really dangerous because the storage can get full which result in unbootable devices.

Is it possible maybe to just check if enough storage is available before downloading a new release?

**Comment by acostach** [comment]
@ClintHaerinck I think you wanted to ping my colleague @pipex for supervisor related items

**Comment by pipex** [comment]
Hi @ClintHaerinck, thanks for the information and sorry for the delay. I think I understand the issue better now, my understanding is that you are probably running into two bugs

- On one side, you are right that setting the env var `BALENA_SUPERVISOR_UPDATE_STRATEGY` has not effect, this is an issue on our backend https://github.com/balena-io/open-balena-api/issues/735 and I'm asking internally why this is the case and we already have a PR for a fix https://github.com/balena-io/open-balena-api/pull/848 
- On the other side, it seems like the `delete-then-download` strategy was not working properly before supervisor [v12.11.10](https://github.com/balena-os/balena-supervisor/blob/master/CHANGELOG.md#v121110), where we introduced a fix. 

Unfortunately balenaOS 2.82.11+rev11 comes with supervisor v12.9.3, prior to that fix, so on OS update you'll see that the image does not get deleted until after the supervisor gets a chance to clean up. 

What you could try is the following. On your device with  balenaOS v2.67.3+rev2, first update the supervisor to the latest version and then run the host OS update. After the update I would expect you should see the images from the unsupported L4T release get deleted before starting the download. 

Could you try that and let us know of the results? Thank you

**Comment by acostach** [comment]
@ClintHaerinck you mentioned 'As result the device gets out of space and unbootable'. Does the issue still happen with the 2.95.8 release of the JN30B? Independent of the supervisor we had https://github.com/balena-os/balena-jetson/issues/271 which seems similar to what you described about booting. This didn't occur on our side but if it happens with your devices it would be good to know if 2.95.8 solves the boot problem.

**Comment by ClintHaerinck** [comment]
> Hi @ClintHaerinck, thanks for the information and sorry for the delay. I think I understand the issue better now, my understanding is that you are probably running into two bugs
> 
>     * On one side, you are right that setting the env var `BALENA_SUPERVISOR_UPDATE_STRATEGY` has not effect, this is an issue on our backend [update strategy and handover timeout not being detect with BALENA_ prefix balena-io/open-balena-api#735](https://github.com/balena-io/open-balena-api/issues/735) and I'm asking internally why this is the case and we already have a PR for a fix [Update config-to-label transform to include `BALENA_*` vars balena-io/open-balena-api#848](https://github.com/balena-io/open-balena-api/pull/848)
> 
>     * On the other side, it seems like the `delete-then-download` strategy was not working properly before supervisor [v12.11.10](https://github.com/balena-os/balena-supervisor/blob/master/CHANGELOG.md#v121110), where we introduced a fix.
> 
> 
> Unfortunately balenaOS 2.82.11+rev11 comes with supervisor v12.9.3, prior to that fix, so on OS update you'll see that the image does not get deleted until after the supervisor gets restarted and performs its initial cleanup.
> 
> What you could try is the following. On your device with balenaOS v2.67.3+rev2, first update the supervisor to the latest version and then run the host OS update. After the update I would expect you should see the images from the unsupported L4T release get deleted before starting the download.
> 
> Could you try that and let us know of the results? Thank you
I tried this method. after the hostOS update the device reboots. When the new hostOS is booted the first time the unsupported L4T releases doesn't get deleted. I need to do an extra reboot to get the behavior you describe.

**Comment by ClintHaerinck** [comment]
@acostach Is it possible to roll-out a BalenaOS with the fix for devices with L4T R32.5.1. Solving the issue looks really important and we still have a lot devices running L4T R32.5.1 and cannot upgrade them to the latest L4T.

**Comment by ClintHaerinck** [comment]
@acostach Or can I just add a script in a container to periodically verify the disk as temporary solution?

**Comment by ClintHaerinck** [comment]
@acostach Ok, thanks.

Just found a device which has following error:
```
root@51964c0:~# sgdisk -v /dev/mmcblk0 2>&1

Warning: There is a gap between the main partition table (ending sector 33)
and the first usable sector (40). This is helpful in some exotic configurations,
but is unusual. The util-linux fdisk program often creates disks like this.
Using 'j' on the experts' menu can adjust this gap.

No problems found. 0 free sectors (0 bytes) available in 0
segments, the largest of which is 0 (0 bytes) in size.
```

Is this another issue? Because the fix you mentioned is checking for "invalid backup GPT header"

**Comment by acostach** [comment]
@ClintHaerinck no, that device should reboot fine as no problems were found by sgdisk.
